web/src/beta/lib/core/Map/useTimelineManager.ts

```diff
@@ -0,0 +1,199 @@
+import { useCallback, useRef, useState, useMemo, MutableRefObject, RefObject } from "react";
+
+import { convertTime } from "@reearth/beta/utils/time";
+
+import { EngineRef } from ".";
+
+type TimelineManager = {
+  readonly timeline: Timeline;
+  readonly options: TimelineOptions;
+  readonly computedTimeline: Timeline;
+  commit: (props: TimelineCommit) => void;
+  onTick: TickEvent;
+  offTick: TickEvent;
+  onCommit: (cb: (committer: TimelineCommitter) => void) => void;
+  offCommit: (cb: (committer: TimelineCommitter) => void) => void;
+  handleTick: (d: Date, clock: { start: Date; stop: Date }) => void;
+  tick: (() => Date | void | undefined) | undefined;
+};
+
+export type TimelineManagerRef = MutableRefObject<TimelineManager | undefined>;
+
+export type Timeline = {
+  current?: Date;
+  start?: Date;
+  stop?: Date;
+};
+
+type TimelineOptions = {
+  animation: boolean;
+  stepType: "rate" | "fixed";
+  multiplier: number;
+  rangeType?: "unbounded" | "clamped" | "bounced";
+};
+
+type TimelineCommand = "PLAY" | "PAUSE" | "SET_TIME" | "SET_OPTIONS";
+
+type TimelineCommit = {
+  cmd: TimelineCommand;
+  payload?:
+    | ({
+        current?: Date | string | undefined;
+        start?: Date | string | undefined;
+        stop?: Date | string | undefined;
+      } & Partial<TimelineOptions>)
+    | undefined;
+  committer: TimelineCommitter;
+};
```

Comment:
I think it would be useful if you declare type by each command like below. Then the usable payload will be more explicit.
```ts
type PlayCommand = {
  cmd: "PLAY",
};
type PauseCommand = {
  cmd: "PAUSE",
};
type SetTimeCommand = {
  cmd: "SET_TIME",
  payload: {
    currentTime: string;
    startTime: string;
    stopTime: string;
  }
};
type TimelineCommit = (PlayCommand | PauseCommand | SetTimeCommand) & { commmiter: TimelineCommiter };
```

Reply:
This is very helpful, updated! Thank you!